proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,580 @@
+`IntraProcessManager`.
+
+![Current IPC Block Diagram](old_ipc.png)
+
+Several shortcomings of the current implementation are listed below.
+
+### Incomplete Quality of Service support
+
+The current implementation can't be used when the QoS durability value is set to `Transient Local`.
+
+The current implementation does not enforce the depth of the QoS history in a correct way. The reason is that
+there is a single ring buffer per `Publisher` and its size is equal to the depth of the `Publisher`'s history.
+A `Publisher` stores a message in the ring buffer and then it sends a meta-message to allow a `Subscription` to retrieve it.
+The `Subscription` correctly stores meta-messages up to the number indicated by its depth of the history, but,
+depending on the frequency at which messages are published and callbacks are triggered, it may happen that a
+meta-message processed from the `Subscription` does not correspond anymore to a valid message in the ring
+buffer, because it has been already overwritten.
+
+More details [here](https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/).
+
+**TODO:** take into account also new QoS: Deadline, Liveliness and Lifespan
+[reference](https://github.com/ros2/design/pull/212).
+
+### Dependent on the RMW
+
+The current implementation of intra-process communication has to send meta-messages from the `Publisher` to
+the `Subscription`s. This is done using the `rmw_publish` function, the implementation of which depends on the
+chosen middleware. This results in the performance of a ROS2 application with intra-process communication
+enabled being heavily dependent on the chosen RMW implementation.
+
+Given the fact that these meta-messages have only to be received from entities within the same process, there
+is space for optimizing how they are transmitted by each RMW.
+However, at the moment none of the supported RMW is actively tackling this issue.
+This results in that the performance of a single process ROS2 application with intra-process communication
+enabled are still worst than what you could expect from a non-ROS application sharing memory between its
+components.
+
+In the following some experimental evidences are quickly presented.
+
+##### Memory requirement
+
+When a `Node` creates a `Publisher` or a `Subscription` to a topic `/MyTopic`, it will also create an
+additional one to the topic `/MyTopic/_intra`. The second topic is the one where meta-messages travel.
+Our [experimental results](https://github.com/irobot-ros/ros2-performance/tree/master/performances/experiments/crystal/pub_sub_memory#adding-more-nodes-x86_64)
+show that creating a `Publisher` or a `Subscription` has a non-negligible memory cost.
+
+Moreover, when using the default RMW implementation, Fast-RTPS, the memory consumed by a ROS2 application
+increases almost exponentially with the number of `Publisher` and `Subscriptions`.
```

Comment:
I think this argument alone is not a good rational for avoiding to use publishers and subscribers. ROS is a publish / subscribe middleware - instead of avoid to use publishers and subscribers the memory consumption should be reduced where possible.

Reply:
No disagreement there. It's more of a practical consideration (than philosophical). I hope you agree that exponential increase in memory with the number of pubs (or subs) is definitely something that can't be ignored. And every core design decision that increases this number makes ROS2 less and less friendly for a resource constrained system (e.g. 32-bit arm linux with 100s of MB of RAM).

This statement (and others about rmw) are simply mentioned as _our_ motivation for creating a new intra-process design.